http/server.ts

```diff
@@ -1,111 +1,117 @@
+  body: Reader;
+  respond: ServerResponder;
+};
 
-/** Continuously read more requests from conn until EOF
- * Calls maybeHandleReq.
- * bufr is empty on a fresh TCP connection.
- * Would be passed around and reused for later request on same conn
- * TODO: make them async function after this change is done
- * https://github.com/tc39/ecma262/pull/1250
- * See https://v8.dev/blog/fast-async
- */
-function serveConn(env: ServeEnv, conn: Conn, bufr?: BufReader) {
-  readRequest(conn, bufr).then(maybeHandleReq.bind(null, env, conn));
-}
+export type ServerResponder = (response: ServerResponse) => Promise<any>;
 
-function maybeHandleReq(env: ServeEnv, conn: Conn, maybeReq: any) {
-  const [req, _err] = maybeReq;
-  if (_err) {
-    conn.close(); // assume EOF for now...
-    return;
-  }
-  env.reqQueue.push(req); // push req to queue
-  env.serveDeferred.resolve(); // signal while loop to process it
+export interface ServerResponse {
+  status?: number;
+  headers?: Headers;
+  body?: Uint8Array | Reader;
 }
 
-export async function* serve(addr: string) {
+export async function* serve(
+  addr: string,
+  cancel: Deferred<void> = defer<void>()
+): AsyncIterableIterator<ServerRequest> {
   const listener = listen("tcp", addr);
-  const env: ServeEnv = {
-    reqQueue: [], // in case multiple promises are ready
-    serveDeferred: deferred()
-  };
-
-  // Routine that keeps calling accept
-  const acceptRoutine = () => {
-    const handleConn = (conn: Conn) => {
-      serveConn(env, conn); // don't block
-      scheduleAccept(); // schedule next accept
-    };
-    const scheduleAccept = () => {
-      listener.accept().then(handleConn);
-    };
-    scheduleAccept();
-  };
-
-  acceptRoutine();
-
-  // Loop hack to allow yield (yield won't work in callbacks)
   while (true) {
-    await env.serveDeferred.promise;
-    env.serveDeferred = deferred(); // use a new deferred
-    let queueToProcess = env.reqQueue;
-    env.reqQueue = [];
-    for (const result of queueToProcess) {
-      yield result;
-      // Continue read more from conn when user is done with the current req
-      // Moving this here makes it easier to manage
-      serveConn(env, result.conn, result.r);
+    const raced = await Promise.race<Conn | void>([
+      listener.accept(),
+      cancel.promise
+    ]);
+    if (!raced) {
+      break;
     }
+    yield readRequest(raced, async response => {
+      await writeResponse(raced, response);
+    });
   }
   listener.close();
 }
 
-export async function listenAndServe(
-  addr: string,
-  handler: (req: ServerRequest) => void
-) {
+export async function listenAndServe(addr: string, handler: HttpHandler) {
   const server = serve(addr);
 
-  for await (const request of server) {
-    await handler(request);
+  for await (const req of server) {
+    await handler(req);
   }
 }
 
-export interface Response {
-  status?: number;
-  headers?: Headers;
-  body?: Uint8Array | Reader;
+export type HttpServerHandler = HttpHandler;
```

Comment:
this type is not used anywhere and it's just an alias, is it needed?

Reply:
No, just forgotten to delete